tests/transport/spdp/spdp_transport.cpp

```diff
@@ -0,0 +1,485 @@
+// Test for the mini-transport in SPDP (sequence number handling)
+
+#include "dds/DCPS/transport/rtps_udp/RtpsUdpInst.h"
+#ifdef ACE_AS_STATIC_LIBS
+#include "dds/DCPS/transport/rtps_udp/RtpsUdp.h"
+#endif
+
+#include "dds/DCPS/transport/framework/NetworkAddress.h"
+
+#include "dds/DCPS/RTPS/BaseMessageTypes.h"
+#include "dds/DCPS/RTPS/GuidGenerator.h"
+#include "dds/DCPS/RTPS/MessageTypes.h"
+#include "dds/DCPS/RTPS/RtpsCoreTypeSupportImpl.h"
+#include "dds/DCPS/RTPS/RtpsDiscovery.h"
+#include "dds/DCPS/RTPS/ParameterListConverter.h"
+#include "dds/DCPS/RTPS/Spdp.h"
+
+#include "dds/DCPS/Service_Participant.h"
+
+#include "ace/Configuration.h"
+#include "ace/Reactor.h"
+#include "ace/Select_Reactor.h"
+
+#include <ace/OS_main.h>
+#include <ace/Thread_Manager.h>
+#include <ace/Reactor.h>
+#include <ace/SOCK_Dgram.h>
+
+#include <exception>
+#include <iostream>
+
+using namespace OpenDDS::DCPS;
+using namespace OpenDDS::RTPS;
+
+// Declared as spdp_friend. Used to Interact Directly with Spdp.
+class DDS_TEST {
+public:
+  DDS_TEST(const RcHandle<Spdp>& spdp, const GUID_t& guid)
+  : spdp_(spdp)
+  , guid_(make_guid(guid.guidPrefix, ENTITYID_PARTICIPANT))
+  {
+  }
+
+  bool check_for_participant(bool expect_participant)
+  {
+    ACE_GUARD_RETURN(ACE_Thread_Mutex, g, spdp_->lock_, true);
+    const bool error = spdp_->has_discovered_participant(guid_) != expect_participant;
+    if (error) {
+      ACE_ERROR((LM_ERROR, ACE_TEXT("Error: expected to %Cfind participant, but did%C\n"),
```

Comment:
I had to google what this meant and it looks like the rule against it is considered obsolete by most guidelines. Regardless, would you prefer "Didn't expect to find the participant, but did" and "Expected to find the participant, but didn't"?

Reply:
I was thinking "expected not to find" would read better than "expected to not find", but perhaps this is obsolete.  Although the succinctness and use of trinary are both good as far as I'm concerned, in terms of readability/maintainability it may be better to just write out both strings, something like `"Error: %C\n", expected ? "message one" : "message two"`

Reply:
I'm actually not picky about the phrasing as along it conveys the meaning, so is this good with you?:

```c++
ACE_ERROR((LM_ERROR, ACE_TEXT("Error: %C\n"), expect_participant ?
  "Expected to find the participant, but didn't" : 
  "Expected not to find the participant, but did");
```

Reply:
Looks good